### PR TITLE
Feat: Responsive tooltips

### PR DESCRIPTION
**Pull request by miksa1987**
Here is my proposition on responsive tooltips, as discussed here: [#115](https://github.com/neherlab/covid19_scenarios/issues/115)

Originally plan was to make a custom component for tooltips, but I had to abandon that as of now as I couldn't get it to work. However, I came up with a kind of responsive y position calculator, and also changed the z-index of tooltips to 999 so other charts do not cover them when hovering. 

Please let me know what you think could be improved, maybe the position calculation could be tweaked a little.

### CODE REVIEW

**Comment by ivan-aksamentov** [comment]
@miksa1987 this looks great!

A non-exhaustive list of things to address:

 - [ ] currently tooltip is sometimes at the bottom even on wider screens. For example, go to Chrome dev tools and click "Toggle device toolbar (Ctrl+Shift+M)". Choose "Laptop with HiDPI screen". This display mode is wide enough, but tooltip is under the graph and is truncated.
  We need to make sure that the feature only activates in single-column mode.  I think that we may use media queries. See example in `FormDatePicker` widget. We need to align breakpoints with bootstrap's breakpoints:
  ```
  $grid-breakpoints: (
    xs: 0,
    sm: 576px,
    md: 768px,
    lg: 992px,
    xl: 1200px
  ) !default;

  ```
(see `node_modules/bootstrap/scss/_variables.scss`)

 - [ ] on mobile the tooltip is truncated by default (most of it is invisible below the screen bottom edge )
 - [ ] could we make tooltip text to arrange in columns responsively?
 - [ ] we need to make tooltip numbers shorter by formatting them for humans. I created https://github.com/neherlab/covid19_scenarios/issues/158
 - [ ] test on different devices, at least in Chrome and Firefox device emulators

It is going to look great. Please continue!

**Comment by miksa1987** [comment]
@ivan-aksamentov I will look deeper into these later in the evening. 

Couple things:

- I would have liked to take advantage of media queries, but the Tooltip component doesn't seem to like classNames, or being wrapped with divs. That's why I had to at least initially implement this the way I did, but I will align it to the breakpoints and investigate if I could  still isolate the responsive Tooltip into its own component.

- If part of the Tooltip is below screen border, should it scroll upon clicking so that all of the Tooltip is visible? Although I'm not completely sure how to do this. Maybe to use refs on charts and always scroll them to the top of screen?

- I think I will also try to tweak the Tooltip top edge calculation so that it would always Be at the very bottom edge of the chart. Will have to see if I can get accurate chart height from somewhere...

- And yes, I will test what it looks like on different screen sizes/devices and browsers. I can provide you with screenshots later If you want?

**Comment by miksa1987** [comment]
@ivan-aksamentov what you think of this? I made sure that Tooltip y position calculation checks screen width from window.innerWidth, and aligned it with the breakpoint into single/double column. So it should show the Tooltip below charts only on single column. The y positioning is somewhat better I think, although not perfect.

Also, I added a functionality to scroll the chart into top of screen when in single column mode, when clicking on the chart. This should address that the Tooltip is always fully visible on mobile, and not blocking the chart.

Here's a couple of screenshots how it should look like on an iPhone X(when clicked on a chart):
mobileview

**Comment by ivan-aksamentov** [comment]
@miksa1987 I love how it scrolls when you click on a chart. This is great user experience.
These are a lot of corner cases probably, but you are doing a fantastic job!

Please rebase and let me know if we can merge. We can fix little kinks after merge too.

**Comment by miksa1987** [comment]
@ivan-aksamentov thank you!

I was intending to add the responsivity to the Tooltip still before I'm totally happy with it though, like discussed here: [#159 ](https://github.com/neherlab/covid19_scenarios/issues/159)

I already started implementing that, and I can probably finish in the evening, then rebase it and make final commit so you can merge.

**Comment by ivan-aksamentov** [comment]
@miksa1987 Okay, there is no rush, take your time. 

Have you seen this PR: https://github.com/neherlab/covid19_scenarios/pull/195 ?
Here is a preview: https://covid19-scenarios-gmm58kuaa.now.sh/

Folks change some desktop styling there quite seriously. I am worried about merge conflicts.
Do you think you can align with them?

They haven't finished yet, but they will hopefully will at some point, so as a side-experiment, maybe you could try on a separate branch to rebase on their changes to see how much of a merge conflict it creates? No need to actually merge.

**Comment by miksa1987** [comment]
@ivan-aksamentov hmm. I'll have to check that out too when I get to my computer. I'm pretty optimistic that there should be no major conflicts since I'm mostly just modifying The Tooltip component, or stuff related to it. Although there might be surprises, so I will see how the rebasing goes.

**Comment by miksa1987** [comment]
@ivan-aksamentov all right! Now I'm pretty happy how it turned out. I finally managed to get the responsive columns just about right. Check it out and if you're happy with it, it can be merged.

File: src/components/Main/Results/AgeBarChart.tsx [hunk 1]
```diff
@@ -4,16 +4,17 @@ import ReactResizeDetector from 'react-resize-detector'
...
 import { useTranslation } from 'react-i18next'
 
-import { Bar, BarChart, CartesianGrid, Legend, Tooltip, XAxis, YAxis, TooltipPayload } from 'recharts'
+import { Bar, BarChart, CartesianGrid, Legend, Tooltip, XAxis, YAxis } from 'recharts'
 
 import { AlgorithmResult } from '../../../algorithms/types/Result.types'
 
 import { SeverityTableRow } from '../Scenario/SeverityTable'
 
-import { numberFormatter } from '../../../helpers/numberFormat'
```
Comment (rcbevans) [inline src/components/Main/Results/AgeBarChart.tsx]: The purpose of the numberFormatter is to localize, but also permit humanizing of displayed values (though this is currently broken in master, for me).

I'm not sure how `t('localized:number', { value: tick })` behaves for formatting, does this abstract the internationalization somehow?

File: src/components/Main/Results/AgeBarChart.tsx [hunk 2]
```diff
@@ -85,19 +81,22 @@ export function AgeBarChart({ showHumanized, data, rates }: SimProps) {
                 }}
               >
                 <XAxis dataKey="name" />
-                <YAxis
-                  label={{ value: t('Cases'), angle: -90, position: 'insideLeft' }}
-                  tickFormatter={tickFormatter}
```
Comment (rcbevans) [inline src/components/Main/Results/AgeBarChart.tsx]: Is this a merge issue?  The YAxis ticks should still be humanized to make them more readable/shorter

File: src/components/Main/Results/AgeBarChart.tsx [hunk 3]
```diff
@@ -85,19 +81,22 @@ export function AgeBarChart({ showHumanized, data, rates }: SimProps) {
...
-                <Legend verticalAlign="top" />
+                <Legend verticalAlign="top"/>
                 <CartesianGrid strokeDasharray="3 3" />
                 <Bar dataKey="peakSevere" fill={colors.severe} name={t('peak severe')} />
                 <Bar dataKey="peakCritical" fill={colors.critical} name={t('peak critical')} />
                 <Bar dataKey="peakOverflow" fill={colors.overflow} name={t('peak overflow')} />
                 <Bar dataKey="totalDead" fill={colors.death} name={t('total deaths')} />
               </BarChart>
+              
+              <div ref={secondRef} />
```
Comment (rcbevans) [inline src/components/Main/Results/AgeBarChart.tsx]: minor: these refs could be slightly more descriptively named

File: src/components/Main/Results/DeterministicLinePlot.tsx [hunk 4]
```diff
@@ -229,8 +224,17 @@ export function DeterministicLinePlot({ data, userResult, logScale, showHumanize
                   domain={[tMin, tMax]}
                   tickCount={7}
                 />
-                <YAxis scale={logScaleString} type="number" domain={[1, 'dataMax']} tickFormatter={yTickFormatter} />
-                <Tooltip formatter={tooltipFormatter} labelFormatter={labelFormatter} />
+                <YAxis
+                  scale={logScaleString}
+                  type="number"
+                  domain={[1, 'dataMax']}
+                  tickFormatter={(tick) => t('localized:number', { value: tick })}
```
Comment (rcbevans) [inline src/components/Main/Results/DeterministicLinePlot.tsx]: This doesn't look like it's accounting for the showHumanized boolean, i.e show as 1.67M rather thatn 1670000

File: src/components/Main/Results/AgeBarChart.tsx (hunk 2 above)
Comment (miksa1987) [reply src/components/Main/Results/AgeBarChart.tsx]: There seems to have been some weirdness related to merging. I'm looking into it.

**Comment by miksa1987** [comment]
@rcbevans @ivan-aksamentov what do you think about it now? I fixed (hopefully all) too eagerly cleaned up lines while merging, and thought of more descriptive names for the refs in AgeBarChart.

**Comment by ivan-aksamentov** [comment]
Related: some more thoughts on improving tooltips: https://github.com/neherlab/covid19_scenarios/issues/233

**Comment by miksa1987** [comment]
@ivan-aksamentov I guess something weird happened while rebasing. I think I'll just create another PR soon, it's probably the easiest way.

**Comment by miksa1987** [comment]
@ivan-aksamentov I created a new PR: [#237](https://github.com/neherlab/covid19_scenarios/pull/237)

So we can ignore this, git went bananas when rebasing.

**Comment by ivan-aksamentov** [comment]
Closing in favor of https://github.com/neherlab/covid19_scenarios/pull/237

**Pull request closed without merging**